armi/reactor/composites.py

```diff
@@ -771,66 +778,80 @@ def _deriveUndefinedVolume(self):
-            )
-            raise RuntimeError("Gluttony Error too many leftovers.")
-        return derivedVolume
+        # Determine the number of `DerivedShape` components and non-derived shape components
+        derivedShapeComps = [c for c in self.getChildren() if c.__class__ is components.DerivedShape]
+        otherComps = [c for c in self.getChildren() if c not in derivedShapeComps]
+
+        otherCompAreas = sum([c.getArea() for c in otherComps])
+        otherCompVolumes = sum([c.getVolume() for c in otherComps])
+
+        remainingArea = self.getMaxArea() - otherCompAreas
+        remainingVolume = self.getMaxVolume() - otherCompVolumes
+
+        geometryError = False
+        # Check for negative area
+        if remainingArea < 0:
+            geometryError = True
+            componentAreas = "\n".join(
+            [
+                f"{c:40s}, {c.getArea():.5e} cm^2"
+                for c in otherComps
+            ])
+            msg = (f"The component areas in {self} exceed the maximum "
+                   f"allowable area based on the geometry. Check that the "
+                   f"geometry is defined correctly.\n"
+                   f"Maximum allowable area: {self.getMaxArea()} cm^2\n"
+                   f"Area of all defined components: {otherCompAreas} cm^2\n"
+                   f"Breakdown: {componentAreas}")
+            runLog.error(msg)
+
+        # Check for negative volume
+        if remainingVolume < 0:
+            geometryError = True
+            componentVols = "\n".join(
+            [
+                f"{c:40s}, {c.getVolume():.5e} cc"
+                for c in otherComps
+            ])
+            msg = (f"The component volumes in {self} exceed the maximum "
+                   f"allowable volume based on the geometry. Check that the "
+                   f"geometry is defined correctly.\n"
+                   f"Maximum allowable volume: {self.getMaxVolume()} cc\n"
+                   f"Volume of all defined components: {otherCompVolumes} cc\n"
+                   f"Breakdown: {componentVols}")
+            runLog.error(msg)
+
+        if geometryError:
+            raise ValueError(f"Negative area/volume errors occurred for {self}. "
+                             "Check log for errors.")
+
+        if len(derivedShapeComps) == 1:
+            c = derivedShapeComps[0]
+
+            # A ZeroDivisonError can occur when the height of the component's
+            # parent is zero, thus the volume may be set within `setVolume`
+            # (using the derived volume calculated here), but the component's
+            # area cannot be determined by directly using this volume. In this
+            # case, the area can be computed by subtracting the area of all
+            # other components from the parent's maximum area.
+            try:
+                c.setVolume(remainingVolume)
+            except ValueError:
+                c.setArea(remainingArea)
```

Comment:
This is a little surprising at this level of abstraction and seems fragile or unpredictable. I think we may be able to make this cleaner by changing the framework behavior at some other level (e.g. within `c.setVolume`). 

We need to make `components` have height regardless of their parent.

Reply:
Alright, we can talk about this more.